src/main/java/net/neoforged/neoforge/transfer/resources/FluidResource.java

```diff
@@ -0,0 +1,336 @@
+     */
+    public static FluidResource of(Fluid fluid) {
+        if (fluid == Fluids.EMPTY) return EMPTY;
+        return fluid.getDefaultResource();
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your fluid is registered
+     * 
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     * @throws IllegalStateException If the underlying default FluidResource has not been yet initialized.
+     */
+    public static FluidResource of(Holder<Fluid> fluid) {
+        return of(fluid.value());
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your fluid is registered
+     * 
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     * @throws IllegalStateException If the underlying default FluidResource when used has not been yet initialized.
+     */
+    public static FluidResource of(Holder<Fluid> fluid, DataComponentPatch patch) {
+        if (fluid.value() == Fluids.EMPTY) return EMPTY;
+        if (patch.isEmpty()) return fluid.value().getDefaultResource();
+        return new FluidResource(new FluidStack(fluid, 1, patch));
+    }
+
+    /**
+     * We wrap a fluid stack which must never be modified.
+     */
+    //This is package private to provide Unsafe access in the scenarios of avoiding allocation
+    // when being used in readonly context. A getInnerStack method could be done, but serves
+    // no functional difference in this case since the field is marked final.
+    final FluidStack innerStack;
+
+    /**
+     * Lazily initialized.
+     */
+    @Nullable
+    private ItemResource filledBucket;
+
+    private FluidResource(FluidStack innerStack) {
+        this.innerStack = innerStack;
+    }
+
+    /**
+     * @return The {@link Fluid} of this resource from the inner {@link FluidStack}
+     */
+    @Override
+    public Fluid getInstanceValue() {
+        return innerStack.getFluid();
+    }
+
+    /**
+     * @return the fluid holder of this resource
+     */
+    @Override
+    public Holder<Fluid> getHolder() {
+        return innerStack.getFluidHolder();
+    }
+
+    /**
+     * Checks if this resource is empty. The resource will be empty if the fluid is {@link Fluids#EMPTY}.
+     *
+     * @return if this resource is empty
+     */
+    @Override
+    public boolean isEmpty() {
+        return innerStack.isEmpty();
+    }
+
+    /**
+     * Returns a copy of this resource with the patch applied.
+     *
+     * @param patch the patch to apply
+     * @return the new resource
+     */
+    @Override
+    public FluidResource withPatch(DataComponentPatch patch) {
```

Comment:
Does it add the patch to the resource or overwrite it? Unclear from the javadoc.